File: solutions/Virtual-Assistant/src/csharp/assistant/Dialogs/Onboarding/OnboardingResponses.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.TemplateManager;
using Microsoft.Bot.Schema;
using VirtualAssistant.Dialogs.Onboarding.Resources;

namespace VirtualAssistant.Dialogs.Onboarding
{
    public class OnboardingResponses : TemplateManager
    {
        private static LanguageTemplateDictionary _responseTemplates = new LanguageTemplateDictionary
        {
            ["default"] = new TemplateIdMap
            {
                {
                    ResponseIds.NamePrompt,
                    (context, data) =>
                    MessageFactory.Text(
                        text: OnboardingStrings.NAME_PROMPT,
                        ssml: OnboardingStrings.NAME_PROMPT,
                        inputHint: InputHints.ExpectingInput)
                },
                {
                    ResponseIds.Greeting,
                    (context, data) =>
                    MessageFactory.Text(
                        text: string.Format(OnboardingStrings.GREETING, data.Name),
                        ssml: string.Format(OnboardingStrings.GREETING, data.Name),
                        inputHint: InputHints.IgnoringInput)
                }
            }
        };

        public OnboardingResponses()
        {
            Register(new DictionaryRenderer(_responseTemplates));
        }

        public class ResponseIds
        {
            public const string NamePrompt = "namePrompt";
            public const string Greeting = "greeting";
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/assistant/src/dialogs/onboarding/onboardingResponses.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import {
    Activity,
    Attachment,
    CardFactory,
    CardImage,
    InputHints,
    MessageFactory,
    TurnContext } from 'botbuilder';
import { __ } from 'i18n';
import {
    DictionaryRenderer,
    LanguageTemplateDictionary,
    TemplateFunction,
    TemplateIdMap} from '../templateManager/dictionaryRenderer';
import { TemplateManager } from '../templateManager/templateManager';

export class OnboardingResponses extends TemplateManager {
    public static responseIds: {
        namePrompt: string;
        locationPrompt: string;
        haveLocation: string;
        addLinkedAccountsMessage: string;
    } = {
        namePrompt: 'namePrompt',
        locationPrompt: 'locationPrompt',
        haveLocation: 'haveLocation',
        addLinkedAccountsMessage: 'linkedAccountsInfo'
    };

    private static readonly responseTemplates: LanguageTemplateDictionary = new Map([
        ['default', <TemplateIdMap> new Map([
            [OnboardingResponses.responseIds.namePrompt, OnboardingResponses.fromResources('onboarding.namePrompt')],
            [OnboardingResponses.responseIds.locationPrompt, OnboardingResponses.fromResources('onboarding.locationPrompt')],
            [OnboardingResponses.responseIds.haveLocation, OnboardingResponses.fromResources('onboarding.haveLocation')],
            [OnboardingResponses.responseIds.addLinkedAccountsMessage,
                // tslint:disable-next-line:no-any
                (context: TurnContext, data: any): Promise<Activity> => OnboardingResponses.buildLinkedAccountsCard(context, data)]
        ])]
   ]);

   // Constructor
    constructor() {
        super();
        this.register(new DictionaryRenderer(OnboardingResponses.responseTemplates));
    }

    // tslint:disable-next-line:no-any
    public static async buildLinkedAccountsCard(turnContext: TurnContext, data: any): Promise<Activity> {
        const title: string = __('onboarding.linkedAccountsInfoTitle');
        const text: string = __('onboarding.linkedAccountsInfoBody');
        const images: (CardImage | string)[] = [
            {
                url: __('onboarding.linkedAccountsInfoUrl'),
                alt: __('onboarding.linkedAccountsInfoAlt')
            }
        ];
        const attachment: Attachment = CardFactory.heroCard(title, text, images);
        const response: Partial<Activity> = MessageFactory.attachment(attachment, title, InputHints.AcceptingInput);

        return Promise.resolve(<Activity> response);
    }

    private static fromResources(name: string): TemplateFunction {
        return (): Promise<string> => Promise.resolve(__(name));
    }

}
